src/utils/offlineRequestHandler.ts

```diff
@@ -0,0 +1,387 @@
+        tx.oncomplete = () => resolve();
+        tx.onerror = () => reject(tx.error);
+      });
+
+      store.add(serializedRequest);
+      await requestPromise;
+      console.log('Request successfully added: ', request, type, stepKey);
+    } catch (error) {
+      console.error('Error adding request to IndexedDB:', error);
+    }
+  }
+
+  // Helper function to serialize request body based on its type
+  private async serializeRequestBody(
+    body: any
+  ): Promise<SerializedRequestBody> {
+    if (body === null) {
+      return { type: 'text', body: '' };
+    }
+    if (body instanceof Blob) {
+      return { type: 'blob', body: await body.arrayBuffer() }; // Convert Blob to ArrayBuffer for storage
+    } else if (body instanceof FormData) {
+      // FormData needs to be converted to a plain object
+      const formDataObj: Record<string, any> = {};
+      body.forEach((value, key) => {
+        formDataObj[key] = value;
+      });
+      return { type: 'formData', body: formDataObj };
+    } else if (body instanceof ArrayBuffer) {
+      return { type: 'arrayBuffer', body: body };
+    } else {
+      // Assume it's text or JSON
+      return { type: 'text', body: await body.text() };
+    }
+  }
+
+  // TODO: what does performance look like while using IndexDB?
+  // Replay requests from IndexedDB
+  public async replayRequests() {
+    if (!navigator.onLine) {
+      // If offline, block until requests are replayed from online event handler
+      await new Promise((resolve) => this._addOnlineSignal(resolve));
+      return;
+    }
+
+    if (this.isReplayingRequests) {
+      return;
+    }
+
+    this.isReplayingRequests = true;
+
+    try {
+      const db = await this.openDatabase();
+      const tx = db.transaction(this.storeName, 'readwrite');
+      const store = tx.objectStore(this.storeName);
+
+      const requestsByStep: Record<string, SerializedRequest[]> = {};
+
+      await new Promise<void>((resolve) => {
+        store.openCursor().onsuccess = (event: Event) => {
+          const cursor = (event.target as IDBRequest<IDBCursorWithValue | null>)
+            .result;
+          if (cursor) {
+            const request = cursor.value as SerializedRequest;
+            if (request.stepKey) {
+              if (!requestsByStep[request.stepKey]) {
+                requestsByStep[request.stepKey] = [];
+              }
+              requestsByStep[request.stepKey].push(request);
+            }
+            cursor.continue();
+          } else {
+            // All entries have been collected, resolve the promise
+            resolve();
+          }
+        };
+      });
+
+      // Replay submitCustom requests immediately
+      const submitCustomRequests = await this.getRequestsByType('submitCustom');
+      await this.replayRequestsInParallel(submitCustomRequests);
+
+      // Replay submitStep requests immediately
+      const submitStepRequests = await this.getRequestsByType('submitStep');
```

Comment:
can we fetch all of the requests from indexdb in the request above so we don't need to create several new transactions afterwards? i think that will improve performance